fix(tiller): rollback deleted release (#3722)

bmperrea opened the pull request:
Solves #3722 by making the changes in #3539 more compatible with the previous behavior.

This gives a recovery option for "oops I deleted my helm release" by allowing rollback, which is intended to be a working feature of helm. Note that purging releases removes the history required to rollback, so this doesn't work in that case. However, purging takes significantly more time, so it's harder to accidentally purge everything.

Signed-off-by: Brent <bmperrea@gmail.com>

mortent commented on pkg/tiller/release_rollback.go:
```diff
@@ -151,11 +152,16 @@ func (s *ReleaseServer) performRollback(currentRelease, targetRelease *release.R
+	// Supersede all previous deployments, see issue #2941.
 	deployed, err := s.env.Releases.DeployedAll(currentRelease.Name)
-	if err != nil {
+	if err != nil && !strings.Contains(err.Error(), "not found") {
```
Is there any way to the call to `DeployedAll` can return an error containing the string "not found"? Seems like the implementation of `DeployedAll` already checks for this condition and returns an error with a different message in this case:
https://github.com/helm/helm/blob/77924fec911168bc490054c67904f1244ab475e3/pkg/storage/storage.go#L152-L154

bmperrea replied:
Great catch. I was thinking it was wrapping the original error. I will update the pr.

bmperrea committed:
fix(tiller): rollback deleted release (#3722)

Solves #3722 by making the changes in #3539 more compatible with the previous behavior.

This gives a recovery option for "oops I deleted my helm release" by allowing rollback, which is intended to be a working feature of helm. Note that purging releases removes the history required to rollback, so this doesn't work in that case. However, purging takes significantly more time, so it's harder to accidentally purge everything.

Signed-off-by: Brent <bmperrea@gmail.com>

bmperrea commented:
@mortent Thanks for the great feedback. Please have another look. Rolling back deleted releases seems to be pretty easy to fix here and quite important so it would be good to get this through.

mortent commented:
This looks good to me, but someone from the core team will need to review and approve.

bacongobbler commented:
tentatively slating for 2.12 for review.

bacongobbler reviewed (approved):
LGTM
